CommonClient.py

```diff
@@ -173,10 +174,69 @@ class CommonContext:
     items_handling: typing.Optional[int] = None
     want_slot_data: bool = True  # should slot_data be retrieved via Connect
 
-    # data package
-    # Contents in flux until connection to server is made, to download correct data for this multiworld.
-    item_names: typing.Dict[int, str] = Utils.KeyedDefaultDict(lambda code: f'Unknown item (ID:{code})')
-    location_names: typing.Dict[int, str] = Utils.KeyedDefaultDict(lambda code: f'Unknown location (ID:{code})')
+    class NameLookupDict:
+        """A specialized dict, with helper methods, for id -> name item/location data package lookups by game."""
+        _game_store: typing.Dict[str, typing.ChainMap[int, str]]
+
+        def __init__(self, ctx: CommonContext, lookup_type: typing.Literal["item", "location"]):
+            self.ctx = ctx
+            self.lookup_type = lookup_type
+            self._unknown_item = lambda key: f"Unknown {lookup_type} (ID: {key})"
+            self._archipelago_lookup = {}
+            self._game_store = collections.defaultdict(
+                lambda: collections.ChainMap(self._archipelago_lookup, Utils.KeyedDefaultDict(self._unknown_item)))
+            self._flat_store = Utils.KeyedDefaultDict(self._unknown_item)
+
+        # noinspection PyTypeChecker
+        def __getitem__(self, key: str) -> typing.Union[typing.ChainMap[int, str], typing.Dict[int, str]]:
+            # TODO: In a future version (0.6.0?) this should be simplified by removing implicit id lookups support.
+            if isinstance(key, int):
+                logger.warning(f"Implicit name lookup by id only is deprecated and only supported to maintain backwards"
+                               f"compatibility for now. If multiple games share the same id for a {self.lookup_type}, "
+                               f"name could be incorrect. Please use `{self.lookup_type}_names.lookup_in_game()` or "
+                               f"`{self.lookup_type}_names.lookup_in_slot()` instead to avoid this issue.")
+                return self._flat_store[key]
+
+            return self._game_store[key]
+
+        def __len__(self):
+            return len(self._game_store)
+
+        def __iter__(self):
+            return iter(self._game_store)
+
+        def __repr__(self):
+            return self._game_store.__repr__()
+
+        def lookup_in_game(self, code: int, game_name: typing.Optional[str] = None) -> str:
+            """Returns the name for an item/location id in the context of a specific game or own game if `game` is
+            omitted.
+            """
+            if game_name is None:
+                game_name = self.ctx.game
+
+            return self._game_store[game_name][code]
```

Comment:
```suggestion
            if game_name is None:
                game_name = self.ctx.game
            assert game_name, f"tried to lookup {self.lookup_type} with no game name available"

            return self._game_store[game_name][code]
```
`CommonContext.game` is `Optional[str]`
so it could still be `None`

An alternative to this is to change the type of `_game_store` to `dict[str | None, ChainMap...]`
Then I think this lookup would result in the "Unknown ID" ? I'm not sure.

Reply:
The only time game (or slot) should be `None` is if the user is disconnected. However, I'll add an assert just to be safe. I'm not going to add `None` as a possible type as that should be only be possible when these are in an invalid state.